Download art for all non-steam game shortcuts

idbrii opened the pull request:
This PR also builds on #8 and #9. 

This is a pretty big PR and adds a feature I'm pretty excited about: entirely automatic art downloads for all of your non-game shortcuts. It tries to figure out the appid and downloads art from steam. If that fails, there's a way to specify backup art (which the itchio integration does but EGS and Xbox do not).

If you added non-steam games manually (from Origin or something we don't yet support), it will also try to get art for them too.

Here's some before and after with a mix of Steam, EGS, Xbox, and itch games:


# Big Picture
![before-after-bigpicture](https://user-images.githubusercontent.com/43559/122524333-dc3a0500-cfcc-11eb-93ec-fe30ed0f49e8.jpg)
# Desktop client
![before-after-desktop](https://user-images.githubusercontent.com/43559/122524335-dcd29b80-cfcc-11eb-9229-17a0f6f68c34.jpg)

\* Prey in those screenshots only works because it's special cased to avoid getting it confused with Prey 2006. It's a good example of how this method isn't foolproof, but I think it's effective enough to be worthwhile.

This PR also adds `--replace-existing` which I found really useful in testing.

Downloading requires the `--download-art` argument since we'd suddenly download a whole lot of stuff which might be bad? Although my steam grid folder is only 14 MB so maybe it's not that bad. Not sure if you want the defaults to be very minimal or not.

If you run steamsync as admin, it creates symlinks instead of making duplicates of art files. That's not documented, but I'm not sure where I'd put it.

jaydenmilne commented on steamsync-library/src/steameditor.py:
```diff
@@ -0,0 +1,334 @@
+            name = stripped.decode().strip()
+            appid = name_to_id.get(name)
+        # Might return None.
+        return appid
```
As an additional fallback/method, it looks like you can just hit the steam search API, eg
```
https://store.steampowered.com/api/storesearch/?term=control&l=english&cc=US
```
returns
```
{
  "total": 9,
  "items": [
    {
      "type": "app",
      "name": "Control Ultimate Edition",
      "id": 870780,
      "price": {
        "currency": "USD",
        "initial": 3999,
        "final": 3999
      },
      "tiny_image": "https://cdn.akamai.steamstatic.com/steam/apps/870780/capsule_231x87.jpg?t=1620649906",
      "metascore": "85",
      "platforms": {
        "windows": true,
        "mac": false,
        "linux": false
      },
      "streamingvideo": false,
      "controller_support": "full"
    },
    {
      "type": "app",
      "name": "Valve Index® Controllers",
      "id": 1059550,
      "price": {
        "currency": "USD",
        "initial": 27900,
        "final": 27900
      },
      "tiny_image": "https://cdn.akamai.steamstatic.com/steam/apps/1059550/capsule_231x87.jpg?t=1623444718",
      "metascore": "",
      "platforms": {
        "windows": true,
        "mac": false,
        "linux": true
      },
      "streamingvideo": false
    }, // etc
  ]
}
```
so you could just grab the first search result and try that if the edit/ Levenshtein distance  isn't too far (https://pypi.org/project/editdistance/0.3.1/)

(be sure to credit https://github.com/MAPReiff/Steam-Game-Search-NPM/blob/master/index.js)

jaydenmilne replied:
This is undocumented so subject to change, etc

idbrii replied:
This seems like a good idea, but I'm not motivated to implement it. I think this PR can standalone without it.

jaydenmilne replied:
Entirely fair

jaydenmilne commented on steamsync-library/src/steameditor.py:
```diff
@@ -0,0 +1,334 @@
+        """
+        page = requests.get(url)
+        if page.status_code == 200:
+            with dest_fname.open('wb') as f:
```
I was getting a crash here because my steam install didn't have the `grid` folder already. You probably need to make that at some point

jaydenmilne commented on steamsync-library/src/steamsync.py:
```diff
@@ -360,21 +345,30 @@ def add_games_to_shortcut_file(steam_path, steamid, games, skip_backup, use_uri)
+        if download_art and exe not in supported_games:
+            # Not a game we defined, but we can use our art pipeline to get
+            # better art.
+            game = GameDefinition(exe, appname, v['appid'], str(Path(exe).parent), '', None, 'otherstore')
```
I think there should be a separate flag for this - I can see how its useful, but I have over 100 shortcuts in my steam library and when I want to just add one it's too long to wait.

jaydenmilne commented on steamsync-library/src/steameditor.py:
```diff
@@ -0,0 +1,334 @@
+        # Always remove hypen because it adds extra uncertainty.
+        name = re_remove_hyphen.sub('', name)
+        # Ignore 'the' vs 'The' differences (fixes Into The Breach)
+        name = name.lower()
```
One other edge case to consider: The game ABZU is stylized in Steam as `ABZÛ` and so it doesn't match an pull down art. Maybe remove diacritics too?

idbrii replied:
There's a [later step that strips non-ascii characters](https://github.com/jaydenmilne/steamsync/pull/10/files/af25af57dd6a4a86a44dbf5dcc5c59217f43a3e5#diff-02da6a6909bff696edefe7bb4ba74adf992b6837ce27be4ff7550027b5c8dd75R169-R173) from game names that catches this for me:

```python
>>> db = SteamDatabase("C:/Program Files (x86)/Steam", os.path.expandvars('$TEMP/steamsync'))
>>> game_name = 'ABZÛ'
>>> appid = db.guess_appid(game_name)
>>> print(game_name, appid)
ABZÛ 384190
```
That resolves to the right store page: https://store.steampowered.com/app/384190/ABZU/

Are you not seeing the same behaviour?

idbrii replied:
Ah, I see. Epic uses "ABZU" and Steam uses "ABZÛ", and since our strip conversion only goes one way, the look up fails. Good catch!

jaydenmilne commented:
Final thoughts for merge:
- This is amazing, you are amazing
- Fix the crash when the grid folder doesn't exist
- I think you should add another flag to try and parse every shortcut. LMK if you have strong opinions on this

Let me know if you want to implement any other heuristics I suggested for guessing steamids

idbrii commented:
Again, this is on top of #9.

I've fixed the crash, supported abzu, actually checked the version number in the app list, expire the app list after a week, and added a flag for add all art. But I didn't try steam search as an additional fallback.

I renamed my grid folder and tested a sample of games that hit a bunch of the boundary conditions.

| Game                      | Source    | images downloaded      |
|---------------------------|-----------|------------------------|
| ABZU                      | epicstore | 4/4                    |
| Control                   | epicstore | 4/4                    |
| Rocket League®            | epicstore | 4/4                    |
| What Remains of Edith Fin | epicstore | 4/4                    |
| Baba Is You               | itchio    | 4/4                    |
| Lenna's Inception         | itchio    | 4/4                    |
| The King's Bird           | itchio    | 4/4                    |
| art of rally              | itchio    | 4/4                    |
| Genesis Noir for Windows  | xbox      | 4/4                    |
| Ghost of a Tale PC        | xbox      | 4/4                    |
| PHOGS!                    | xbox      | 4/4                    |
| ReCore                    | xbox      | 2/4 (old game)         |
| Tetris® Effect: Connected | xbox      | 0/4 (Appid not found.) |

Tetris Effect doesn't work because weirdly the game's appid is listed as `"omega": 1003590,` in the downloaded json. I swear this worked earlier, so maybe they changed something since it's still not out on Steam yet. I guess ReCore is so old and unpopular that they didn't bother to upload new steam grid art for it.

Then I ran with `--download-art-all-shortcuts` and it looks good, but I didn't check as closely. (Downloaded new art for 35/37 games!)

jaydenmilne reviewed (approved):
Fantastic work

idbrii commented:
Oh, I also tested downloading art for a UWPHook shortcut (Spiritfarer) and that works well too.

👍🏽
